File: android/totvs-camera-vision/src/main/java/com/totvs/camera/vision/face/FaceDetector.kt
package com.totvs.camera.vision.face

import android.graphics.PointF
import android.util.Log
import android.util.Size
import androidx.core.graphics.toRectF
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceDetectorOptions.*
import com.google.mlkit.vision.face.FaceLandmark
import com.totvs.camera.core.ImageProxy
import com.totvs.camera.vision.AbstractVisionDetector
import com.totvs.camera.vision.VisionDetector
import com.totvs.camera.vision.core.SelectionStrategy
import com.totvs.camera.vision.core.VisionModuleOptions.DEBUG_ENABLED
import com.totvs.camera.vision.utils.exclusiveUse
import java.util.concurrent.Executor

/**
 * Detector dedicated to identity faces. This detector is a _Single emission_ detector.
 *
 * This detector is focused in classifications and landmark detections, following the
 * recommendations on:
 * https://developers.google.com/ml-kit/vision/face-detection/android#performance_tips
 *
 * If a contour or another kind of detector is required, this detector offers an interface
 * for subclasses to customize the detector capabilities.
 *
 * As-is this detector keeps a singleton of a high-accuracy detector.
 *
 * This detector relies on the new Firebase detection API.
 *
 * @see [VisionDetector]
 */
open class FaceDetector(
    private val selectFace: SelectionStrategy<Face> = FIRST
) : AbstractVisionDetector<FaceObject>(FaceDetector) {

    private val detector by lazy { FaceDetection.getClient(getDetectorOptions()) }

    override fun detect(
        executor: Executor,
        image: ImageProxy,
        onDetected: (FaceObject) -> Unit
    ) {
        if (null == image.image) {
            return onDetected(NullFaceObject)
        }
        // we require to use this image exclusively and nobody else can read the data until
        // we're done with it.
        val inputImage = image.exclusiveUse {
            InputImage.fromMediaImage(
                it.image!!,
                it.imageInfo.rotationDegrees
            )
        }

        // we perform manual executor execution instead of passing the executor to Firebase
        // because if the executor is shut down before this callback is called,
        // Firebase will popup the exception, here instead we log it
        detector.process(inputImage)
            .addOnSuccessListener { faces ->
                // to chose the best face.
                executor.executeCatching(onDetected) {
                    onDetected(
                        if (faces.isEmpty()) NullFaceObject else mapToFaceObject(
                            selectFace(faces),
                            image.imageInfo.rotationDegrees,
                            Size(image.width, image.height)
                        )
                    )
                }
            }
            .addOnFailureListener {
                executor.executeCatching(onDetected) { onDetected(NullFaceObject) }
            }
    }

    /**
     * Readable name
     */
    override fun toString() = TAG

    /**
     * Get the detector options used for this instance of the face detector.
     */
    open fun getDetectorOptions(): FaceDetectorOptions = fastModeOptions()

    /**
     * Map the MLKIT vision face to face object
     */
    open fun mapToFaceObject(
        face: Face,
        rotation: Int,
        sourceSize: Size
    ) = FaceObject(
        sourceSize = sourceSize,
        boundingBox = face.boundingBox.toRectF(),
        sourceRotationDegrees = rotation,
        width = face.boundingBox.width().toFloat(),
        height = face.boundingBox.height().toFloat(),
        eyesOpenProbability = EyesOpenProbability(
            face.leftEyeOpenProbability   ?: 0.0f,
            face.rightEyeOpenProbability ?: 0.0f
        ),
        eulerZ = face.headEulerAngleZ,
        eulerY = face.headEulerAngleY,
        landmarks = extractLandmarks(face)
    )

    /**
     * Extract all the recognized landmarks. We consider a landmark as recognized
     * if there's a corespondent type [Landmark] for it. If there isn't then the landmark is
     * ignored
     */
    protected fun extractLandmarks(face: Face): List<Landmark> {
        val landmarks = mutableListOf<Landmark>()
        // left eye
        face.getLandmark(FaceLandmark.LEFT_EYE)?.let {
            landmarks.add(LeftEye(PointF(it.position.x, it.position.y)))
        }
        // right eye
        face.getLandmark(FaceLandmark.RIGHT_EYE)?.let {
            landmarks.add(RightEye(PointF(it.position.x, it.position.y)))
        }
        face.getLandmark(FaceLandmark.NOSE_BASE)?.let {
            landmarks.add(Nose(PointF(it.position.x, it.position.y)))
        }
        return landmarks
    }

    /**
     * Utility method to run safely on the executor a blocks
     */
    protected fun Executor.executeCatching(
        onDetected: (FaceObject) -> Unit,
        block: () -> Unit
    ) = execute {
        runCatching(block).exceptionOrNull()?.let {
            if (DEBUG_ENABLED) {
                Log.e(TAG, "", it)
            }
            onDetected(NullFaceObject)
        }
    }

    companion object : VisionDetector.Key<FaceDetector> {

        private const val TAG = "FaceDetector"

        private fun fastModeOptions(): FaceDetectorOptions = Builder()
            .setClassificationMode(CLASSIFICATION_MODE_ALL)
            .setLandmarkMode(LANDMARK_MODE_ALL)
            .setContourMode(CONTOUR_MODE_NONE)
            .setPerformanceMode(PERFORMANCE_MODE_FAST)
            .build()

        /**
         * Strategy for selecting the first face
         */
        val FIRST = object : SelectionStrategy<Face> {
            override fun invoke(faces: List<Face>): Face =
                faces.first()
        }
    }
}

File: ios/TOTVSCameraKit/Vision/Face/FaceDetector.swift
//
//  FaceDetector.swift
//  TOTVSCameraKit
//
//  Created by Jansel Rodriguez on 7/11/20.
//  Copyright © 2020 Jansel. All rights reserved.
//

import Foundation
import MLKitVision
import MLKitFaceDetection

fileprivate typealias MLKitFaceDetector = MLKitFaceDetection.FaceDetector

/**
* Detector dedicated to identity faces. This detector is a _Single emission_ detector.
*
* This detector is focused in classifications and landmark detections, following the
* recommendations on:
* https://developers.google.com/ml-kit/vision/face-detection/android#performance_tips
*
* If a contour or another kind of detector is required, this detector offers an interface
* for subclasses to customize the detector capabilities.
*
* As-is this detector keeps a singleton of a high-accuracy detector.
*
* This detector relies on the new Firebase detection API.
*
* @see [VisionDetector]
*/
open class FaceDetector : VisionDetector {
    
    public static let `default` = FaceDetector()
    
    public static var key: AnyHashable { DETECTOR_KEY }
    
    private let selectFace: SelectionStrategy<Face> = SelectMostProminent
    
    private var detectorOptions: FaceDetectorOptions {
        let options = FaceDetectorOptions()
        options.performanceMode = .accurate
        options.contourMode = .none
        options.landmarkMode = .all
        options.classificationMode = .all
        
        return options
    }
    
    private lazy var detector = MLKitFaceDetector.faceDetector(options: detectorOptions)
    
    private init() { }
    
    public func detect(on queue: DispatchQueue, image: ImageProxy, onDetected: (VisionObject) -> Void) {
        guard let buffer = image.buffer else {
            onDetected(NullFaceObject)
            return
        }
        
        let visionImage = VisionImage(buffer: buffer)
        visionImage.orientation = imageOrientation(
            facing: image.imageInfo.sourceFacing,
            imageOrientation: image.imageInfo.orientation
        )
        
        var faces = [Face]()
        do {
            faces = try detector.results(in: visionImage)
        } catch let error {
            print("FaceDetector for error at detecting faces: \(error)")
        }
                
        guard !faces.isEmpty, let face = selectFace(faces) else {
            onDetected(NullFaceObject)
            return
        }
        onDetected(mapToFaceObject(
            face: face,
            imageWidth: image.width,
            imageHeight: image.height
        ))
    }
    
    /**
     * Map the MLKIT vision face to face object
     */
    private func mapToFaceObject(face: Face, imageWidth: Int, imageHeight: Int) -> FaceObject {
        return FaceObject(
            width: face.frame.width,
            height: face.frame.height,
            eyesOpenProbability: EyesOpenProbability(
                left: face.leftEyeOpenProbability,
                right: face.rightEyeOpenProbability
            ),
            sourceSize: CGSize(width: imageWidth, height: imageHeight),
            boundingBox: face.frame,
            landmarks: extractLandmarks(from: face)
        )
    }
    
    /**
     * Extract all the recognized landmarks. We consider a landmark as recognized
     * if there's a corespondent type [Landmark] for it. If there isn't then the landmark is
     * ignored
     */
    private func extractLandmarks(from face: Face) -> [Landmark] {
        var landmarks = [Landmark]()
        
        if let l = face.landmark(ofType: .leftEye) {
            landmarks.append(Landmark(name: .leftEye, position: l.position.toPoint))
        }
        if let r = face.landmark(ofType: .rightEye) {
            landmarks.append(Landmark(name: .rightEye, position: r.position.toPoint))
        }
        if let n = face.landmark(ofType: .noseBase) {
            landmarks.append(Landmark(name: .nose, position: n.position.toPoint))
        }
        return landmarks
    }
    
    /// Get the right orientation contained in a CMSampleBuffer image
    private func imageOrientation(facing: CameraFacing, imageOrientation: UIDeviceOrientation) -> UIImage.Orientation {
        switch imageOrientation {
        case .portrait:
            return facing == .front ? .leftMirrored : .right
        case .landscapeLeft:
            return facing == .front ? .downMirrored : .up
        case .portraitUpsideDown:
            return facing == .front ? .rightMirrored : .left
        case .landscapeRight:
            return facing == .front ? .upMirrored : .down
        case .faceDown, .faceUp, .unknown:
            return .up
        @unknown default: fatalError()
        }
    }
}

fileprivate extension VisionPoint {
    var toPoint: CGPoint { CGPoint(x: x, y: y) }
}

fileprivate let DETECTOR_KEY = 130290

/**
 * Strategy for selecting the most prominent face
 */
fileprivate let SelectMostProminent : SelectionStrategy<Face> = { (objects: Array<Face>) in objects.first }
